source/credentials_provider_ecs.c

```diff
@@ -629,3 +628,93 @@ struct aws_credentials_provider *aws_credentials_provider_new_ecs(
 
     return NULL;
 }
+
+struct aws_credentials_provider *aws_credentials_provider_new_ecs_from_environment(
+    struct aws_allocator *allocator,
+    const struct aws_credentials_provider_ecs_environment_options *options) {
+
+    struct aws_credentials_provider_ecs_options explicit_options = {
+        .shutdown_options = options->shutdown_options,
+        .bootstrap = options->bootstrap,
+        .function_table = options->function_table,
+    };
+
+    struct aws_string *relative_uri_str = NULL;
+    struct aws_string *full_uri_str = NULL;
+    struct aws_uri full_uri;
+    AWS_ZERO_STRUCT(full_uri);
+    struct aws_credentials_provider *provider = NULL;
+
+    /* Read the auth token variables */
+    struct aws_string *ecs_env_token_file_path = NULL;
+    struct aws_string *ecs_env_token = NULL;
+    if (aws_get_environment_value(allocator, s_ecs_creds_env_token_file, &ecs_env_token_file_path) != AWS_OP_SUCCESS ||
+        aws_get_environment_value(allocator, s_ecs_creds_env_token, &ecs_env_token) != AWS_OP_SUCCESS) {
+        goto cleanup;
+    }
+    if (ecs_env_token_file_path != NULL) {
+        explicit_options.auth_token_file_path = aws_byte_cursor_from_string(ecs_env_token_file_path);
+    }
+    if (ecs_env_token != NULL) {
+        explicit_options.auth_token = aws_byte_cursor_from_string(ecs_env_token);
+    }
+
+    if (aws_get_environment_value(allocator, s_ecs_creds_env_relative_uri, &relative_uri_str) == AWS_OP_SUCCESS &&
+        relative_uri_str != NULL && relative_uri_str->len != 0) {
+
+        /* Using RELATIVE_URI */
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "ECS provider: using relative uri %s",
+            aws_string_c_str(relative_uri_str));
+
+        explicit_options.path_and_query = aws_byte_cursor_from_string(relative_uri_str);
+        explicit_options.host = aws_byte_cursor_from_string(s_ecs_host);
+        explicit_options.port = 80;
+
+        provider = aws_credentials_provider_new_ecs(allocator, &explicit_options);
+
+    } else if (
+        aws_get_environment_value(allocator, s_ecs_creds_env_full_uri, &full_uri_str) == AWS_OP_SUCCESS &&
+        full_uri_str != NULL && full_uri_str->len != 0) {
+
+        /* Using FULL_URI */
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "ECS provider: using full uri %s", aws_string_c_str(full_uri_str));
+        struct aws_byte_cursor full_uri_cursor = aws_byte_cursor_from_string(full_uri_str);
+        if (aws_uri_init_parse(&full_uri, allocator, &full_uri_cursor)) {
+            goto cleanup;
```

Comment:
Log about how AWS_CONTAINER_CREDENTIALS_FULL_URI was invalid. Someone hitting this error and relying on the default chain would have a hell of a time figuring that out

Reply:
debatable: aws_uri_init_parse() failure raises `AWS_ERROR_MALFORMED_INPUT_STRING`

would this be better as AWS_AUTH_CREDENTIALS_PROVIDER_INVALID_ENVIRONMENT? something else?

Reply:
I have added a log, I think `AWS_ERROR_MALFORMED_INPUT_STRING` is fine since this will only be raised if someone has an invalid uri.